frontend/tests/oauth.test.ts

```diff
@@ -0,0 +1,113 @@
+﻿import {type APIRequestContext, type Browser, expect} from '@playwright/test';
+import {test} from './fixtures';
+import {LoginPage} from './pages/loginPage';
+import {OidcDebuggerPage} from './pages/oidcDebuggerPage';
+import {addUserToProject, loginAs, logout, preApproveOauthApp} from './utils/authHelpers';
+import {OauthApprovalPage} from './pages/oauthApprovalPage';
+import {elawaProjectId, serverBaseUrl} from './envVars';
+
+// we've got a matrix of flows to test
+// pre approved, not approved yet,
+// logged in already, not logged in
+
+//use https://oidcdebugger.com/ to test oauth
+const flows = [
+  {approved: true, loggedIn: false},
+  {approved: false, loggedIn: false},
+  {approved: true, loggedIn: true},
+  {approved: false, loggedIn: true}
+];
+
+test.describe('oauth tests', () => {
+  for (const flow of flows) {
+    const name = `can login with oauth, ${flow.approved ? 'pre-approved' : 'not approved'}, ${flow.loggedIn ? 'logged in' : 'not logged in'}`;
+    test(name, async ({page, tempUser}) => {
+
+      if (flow.approved || flow.loggedIn) await loginAs(page.request, tempUser);
+      if (flow.approved) {
+        await preApproveOauthApp(page.request, OidcDebuggerPage.clientId, OidcDebuggerPage.scopes);
+      }
+      if (!flow.loggedIn) {
+        await logout(page);
+      }
+
+      const oauthTestPage = await new OidcDebuggerPage(page).goto();
+      await oauthTestPage.fillForm(serverBaseUrl);
+      await oauthTestPage.submit();
+      const approvalPage = new OauthApprovalPage(page);
+
+      if (!flow.loggedIn) {
+        await expect(page).not.toHaveURL(OidcDebuggerPage.redirectUrl);
+        await approvalPage.expectNotOnPage();
+
+        const loginPage = await new LoginPage(page).waitFor();
+        await loginPage.fillForm(tempUser.email, tempUser.password);
+        await loginPage.submit();
+      }
+      if (!flow.approved) {
+        await expect(page).not.toHaveURL(OidcDebuggerPage.redirectUrl);
+
+        await approvalPage.waitFor();
+        await approvalPage.clickAuthorize();
+      }
+
+      await oauthTestPage.waitForDebuggerPage();
+      const result = await oauthTestPage.getDebuggerIdToken();
+      expect(result).not.toBeFalsy();
+    });
+  }
+
+  async function addUserToProjectNewContext(browser: Browser, userId: string) {
```

Comment:
I very much like how you use a new context. We talked about this in the office and how it would have been better if we had always done this for this sort of task.

Because `newContext` is sort of "name metadata" maybe it would be appropriate to seperate it/make that more explicit?

In a cleaner code base, we would likely just drop it, because we never don't use a new context. 😔

```suggestion
  async function addUserToProject_newContext(browser: Browser, userId: string) {
```

Reply:
Yeah, I could see is making this much better, imagine a fixture which was incjected and you could do this: `api.AddUserToProject` to use the current context, or `api.asAdmin.AddUserToProject` which we use a context where you're already logged in as an admin, would speed up tests a lot too if it only needed to login once. I was going to just change the name but eslint got mad with the name as it doesn't match our rules.